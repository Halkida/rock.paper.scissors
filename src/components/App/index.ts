import withAuthCheck from '@/hocs/with-auth-check';
import withOAuthCheck from '@/hocs/with-oauth-check';
import { compose } from 'redux';
import { App } from './App';

export default compose(
  withOAuthCheck,
  withAuthCheck,
)(App);
