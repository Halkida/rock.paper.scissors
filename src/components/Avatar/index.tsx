export { AvatarImg } from './AvatarImg';
export { AvatarModal } from './AvatarModal';