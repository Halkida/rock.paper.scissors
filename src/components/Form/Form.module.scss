@import '@/styles/variables.scss';

.form {
  display: flex;
  flex-direction: column;
  min-width: 350px;
  margin-bottom: 20px;

  &__title {
    color: $color-white;
    font-size: $font-medium;
  }

  &__body {
    display: flex;
    flex-direction: column;
  }
}