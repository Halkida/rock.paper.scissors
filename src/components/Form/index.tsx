export { Form } from './Form';
export { useForm, FieldError } from './useForm';