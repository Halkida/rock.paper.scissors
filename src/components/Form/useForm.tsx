import { useState, FormEvent } from 'react';

export const useForm = <T extends Record<keyof T, any> = Record<string, unknown>>(options: any) => {
  const [data, setData] = useState(options?.initialValues || {});
  const [errors, setErrors] = useState({});

  const handleChange = (key: string) => (value: string) => {
    setData((prevData: Record<string, unknown>) => {
      return { ...prevData, [key]: value };
    });
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const validationConfig = options?.validationConfig;
    let isValid = true;
    if (validationConfig) {
      const errors: Record<string, string> = {};
      for (const key in validationConfig) {
        const value = typeof data[key] === 'undefined' ? '' : data[key];
        const validation = validationConfig[key];

        const pattern = validation?.pattern;
        if (pattern?.value && !value.match(pattern.value)) {
          isValid = false;
          errors[key] = pattern.message;
        }

        const custom = validation?.custom;
        if (custom?.isValid && !custom.isValid(value)) {
          isValid = false;
          errors[key] = custom.message;
        }
      }

      setErrors(errors);
    }

    if (isValid && options?.onSubmit) {
      options.onSubmit(data);
    }
  };

  return {
    data,
    handleChange,
    handleSubmit,
    errors
  };
};