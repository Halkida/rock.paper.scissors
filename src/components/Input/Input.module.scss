@import '@/styles/variables.scss';

.input {
  background: transparent;
  border: 2px solid $color-shark;
  color: $color-tulip-tree;

  width: 100%;
  padding: 5px;
  height: 40px;
  border-radius: $default-border-radius;
  margin: 5px 0;
  outline: none;

  &_invalid {
    border-color: $color-geraldine;
  }

  &:disabled {
    background-color: $color-silver;
    pointer-events: none;
  }

  &[type=number]::-webkit-inner-spin-button,
  &[type=number]::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  &[type=number] {
    -moz-appearance: textfield;
  }
}

.error-text {
  font-size: $font-small;
  color: $color-geraldine;
}