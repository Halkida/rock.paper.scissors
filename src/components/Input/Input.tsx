import { FC, InputHTMLAttributes, ChangeEvent, useCallback } from 'react';
import classNames from 'classnames/bind';
import styles from'./Input.module.scss';

type OwnProps = {
  isValid?: boolean;
  errorText?: string;
  type?: 'number' | 'text' | 'password',
  onChange?: (value: string) => void
};

type ReactInputAttributes = InputHTMLAttributes<HTMLInputElement>;
type InputProps = OwnProps & Omit<ReactInputAttributes, keyof OwnProps>
type Props = FC<InputProps>;

const cx = classNames.bind(styles);

export const Input: Props = function Input(
  { isValid = true,
    errorText = '',
    type = 'text',
    onChange,
    value,
    ...InputHTMLAttributes }
  ) {

  const className = cx({
    ['input']: true,
    ['input_invalid']: !isValid
  });

  const changeHandler = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    if (onChange) {
      onChange(value);
    }
  }, [value]);

  return (
    <>
      <input onChange={ changeHandler }
      type={type} className={className} { ...InputHTMLAttributes } />
      {
        !isValid && errorText &&
        <span className={styles['error-text']} >{ errorText }</span>
      }
    </>
  );
};
