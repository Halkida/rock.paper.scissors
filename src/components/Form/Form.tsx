import { FC, FormEvent } from 'react';
import classNames from 'classnames/bind';
import { Button } from '@/components/Button';
import styles from'./Form.module.scss';

type OwnFormProps = {
  className?: string,
  onSubmit: (event: FormEvent<HTMLFormElement>) => void;
  renderFields: () => JSX.Element,
  submitText?: string
  title?: string
};
type FormProps = FC<OwnFormProps>;

const cx = classNames.bind(styles);

export const Form: FormProps = function Form(
  { onSubmit,
    renderFields,
    title,
    submitText = 'Отправить',
    className = '',
  }) {
  const formClassNames = cx({
    [className]: true,
    ['form']: true,
  })
  return (
    <div className={ formClassNames }>
      { title &&
        <span className={ styles.form__title }>
          {title}
        </span>
      }
      <form onSubmit={onSubmit} action='' className={styles.form__body}>
        { renderFields() }
        <Button type='submit' size='small'>{submitText}</Button>
      </form>
    </div>
  );
};