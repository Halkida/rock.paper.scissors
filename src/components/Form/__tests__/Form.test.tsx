import { render, fireEvent } from '@testing-library/react';
import * as renderer from 'react-test-renderer';
import { Form } from '@/components/Form';
import { Input } from '@/components/Input';

describe('Form component testing', () => {
  const loginValue = 'login';
  const passwordValue = 'pass';
  const formFields = (
    <>
      <Input
        value={loginValue}
        name='login'
        placeholder='Login placeholder'
        isValid={true}
      />
      <Input
        value={passwordValue}
        name='password'
        placeholder='Password placeholder'
        isValid={true}
      />
    </>
  );

  test('Form snapshot test', () => {
    const component = renderer.create(
      <Form
        onSubmit={() => {}}
        title='Test form'
        submitText='Submit text'
        renderFields={() => formFields}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Form submit test', () => {
    const onSubmit = jest.fn(e => e.preventDefault());
    const { getByText } = render(
      <Form
        onSubmit={onSubmit}
        title='Test form'
        submitText='Submit button text'
        renderFields={() => formFields}
      />
    );
    const submitButton = getByText('Submit button text');
    fireEvent.click(submitButton);

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});