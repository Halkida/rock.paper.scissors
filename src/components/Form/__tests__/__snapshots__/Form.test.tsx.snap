// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component testing Form snapshot test 1`] = `
<div>
  <span>
    Test form
  </span>
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      className="input"
      name="login"
      onChange={[Function]}
      placeholder="Login placeholder"
      type="text"
    />
    <input
      className="input"
      name="password"
      onChange={[Function]}
      placeholder="Password placeholder"
      type="text"
    />
    <button
      className=" button button_default button_small"
      type="submit"
    >
      Submit text
    </button>
  </form>
</div>
`;
